### PR TITLE
fix(chain): handle approvals properly

### PR DESCRIPTION
**Pull request by bowenwang1996**
Address a couple issues related to approvals:
* If we do not know about the parent block that the approval points to, we should still perform some validity check before putting it in the cache.
* If the signature of an approval is invalid, it should be dropped.
* Do not send approvals if the node is not validating or is about to become a validator. Fixes #3015 

Test plan
----------
* `test_invalid_approvals`

### CODE REVIEW

File: chain/client/src/client.rs [hunk 1]
```diff
@@ -1041,30 +1103,57 @@ impl Client {
...
-        };
-
+        //        let is_validator =
+        //            |epoch_id, block_hash, account_id, runtime_adapter: &Arc<dyn RuntimeAdapter>| {
+        //                match runtime_adapter.get_validator_by_account_id(epoch_id, block_hash, account_id)
+        //                {
+        //                    Ok((_, is_slashed)) => !is_slashed,
+        //                    Err(_) => false,
+        //                }
+        //            };
```
Comment (frol) [inline chain/client/src/client.rs:1118]: Is this expected to be commented out?

**Review by frol — APPROVED** [review]
I was pinged for review due to the changes in the primitives, those LGTM, I like the switch from a boolean flag to an explicit enum type.

File: chain/client/src/client.rs [hunk 2]
```diff
@@ -1111,6 +1161,10 @@ impl Client {
                 signature,
             ) {
                 Ok(true) => {}
+                Ok(false) => self.network_adapter.do_send(NetworkRequests::BanPeer {
```
[the remark this answers is not in the document]
Comment (bowenwang1996) [reply chain/client/src/client.rs:1164]: I thought the peer id passed here is the author of the message, but apparently I was wrong. Fixed

**Review by SkidanovAlex — APPROVED** [review]

**Comment by SkidanovAlex** [comment]
I would heavily recommend removing the `BanPeer`.
Reasons:
1. It only works for direct peers. Thus, it cannot be a full solution. If someone wants to execute an attack, they will easily bypass it by using the gossip network. Thus, this logic serves no purpose.
2. Anything that involves `BanPeer`, if has a bug, can be used to eclipse a node. Having some logic that on itself serves no purpose and doesn't address any possible attack, but, if wrong, can be used to execute eclipse attacks, seems wrong.
3. It can also make one analysing the rest of the logic incorrectly assume that one can only send an invalid approval a limited number of times, and thus reason about the rest of the code incorrectly.

**Pull request merged**